### PR TITLE
[train] add XGBoostTrainer user guide

### PR DESCRIPTION
**Pull request by matthewdeng**
## Why are these changes needed?

This PR adds a comprehensive guide for distributed XGBoost training with Ray Train. The guide follows the same structure as our other framework-specific guides.

The PR also adds a deprecation notice to the older XGBoost/LightGBM notebook and includes the guide in the main Train documentation navigation.

In a follow-up PR, we will create a similar guide for LightGBM.

## Checks

https://anyscale-ray--52355.com.readthedocs.build/en/52355/train/getting-started-xgboost.html

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by chris-ray-zhang — APPROVED** [review]
great work @matthewdeng !

File: doc/source/train/getting-started-xgboost.rst [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+                model_path = os.path.join(checkpoint_dir, RayTrainReportCallback.CHECKPOINT_NAME)
+                model = xgboost.Booster()
+                model.load_model(model_path)
+
+
+Set up a training function
+--------------------------
+
+First, update your training code to support distributed training.
+Begin by wrapping your code in a :ref:`training function <train-overview-training-function>`:
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:159]: Can you add a mention that the Scikit-learn estimator interface also works here?

Link to this: https://xgboost.readthedocs.io/en/latest/python/sklearn_estimator.html

File: doc/source/train/getting-started-xgboost.rst [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+        ...
+
+    config = {"label_column": "y", "num_boost_round": 10}
+    trainer = ray.train.xgboost.XGBoostTrainer(train_func, train_loop_config=config, ...)
+
+.. warning::
+
+    Avoid passing large data objects through `train_loop_config` to reduce the
+    serialization and deserialization overhead. Instead, it's preferred to
+    initialize large objects (e.g. datasets, models) directly in `train_func`.
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:187]:
```suggestion
    serialization and deserialization overhead. Instead,
    initialize large objects (e.g. datasets, models) directly in `train_func`.
```

File: doc/source/train/getting-started-xgboost.rst [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+            callbacks=[
+                RayTrainReportCallback(
+                    metrics={"loss": "eval-logloss"}, frequency=1
+                )
+            ],
+        )
+        ...
+
+
+Reporting metrics and checkpoints to Ray Train enables integration with Ray Tune and :ref:`fault-tolerant training <train-fault-tolerance>`.
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:238]: nit: sounds a little bit like Ray Tune fault tolerant training right now

```suggestion
Reporting metrics and checkpoints to Ray Train enables :ref:`fault-tolerant training <train-fault-tolerance>` and the integration with Ray Tune.
```

File: doc/source/train/getting-started-xgboost.rst [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+        -    data = config["data"]
+        -    model = config["model"]
+
+        +    data = load_dataset()
+        +    model = load_model()
+             ...
+
+         trainer = ray.train.xgboost.XGBoostTrainer(train_func, train_loop_config=config, ...)
+
+Ray Train automatically sets up the Rabit communicator for XGBoost, which handles the distributed communication between workers.
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:211]:
```suggestion
Ray Train automatically performs the worker communication setup that is needed to do distributed xgboost training.
```

File: doc/source/train/getting-started-xgboost.rst [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+
+Report metrics and save checkpoints
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To persist your checkpoints and monitor training progress, add a
+:class:`ray.train.xgboost.RayTrainReportCallback` utility callback to your Trainer:
+
+
+.. code-block:: diff
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:220]:
```suggestion
.. code-block:: python
```

File: doc/source/train/getting-started-xgboost.rst [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+     import xgboost
+     from ray.train.xgboost import RayTrainReportCallback
+
+     def train_func():
+        ...
+        bst = xgboost.train(
+            ...,
+            callbacks=[
+                RayTrainReportCallback(
+                    metrics={"loss": "eval-logloss"}, frequency=1
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:231]: the dict renaming map thing is a bit advanced/confusing. let's just keep the metric names that are returned by xgboost.

```suggestion
                RayTrainReportCallback(
                    metrics=["eval-logloss"], frequency=1
```

File: doc/source/train/getting-started-xgboost.rst [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+        )
+        ...
+
+
+Reporting metrics and checkpoints to Ray Train enables integration with Ray Tune and :ref:`fault-tolerant training <train-fault-tolerance>`.
+
+Loading data
+------------
+
+When running distributed XGBoost training, you will need each worker to access a different shard of the dataset.
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:243]:
```suggestion
When running distributed XGBoost training, each worker should use a different shard of the dataset.
```

File: doc/source/train/getting-started-xgboost.rst [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+
+    def train_func():
+        rank = ray.train.get_world_rank()
+        dtrain = get_train_dataset(rank)
+        deval = get_eval_dataset(rank)
+        ...
+
+A common way to do this is to pre-shard the dataset, so each worker reads from a different partition of the dataset. 
+
+For more flexibility, Ray Data provides a solution for sharding the dataset at runtime.
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:265]:
```suggestion
A common way to do this is to pre-shard the dataset and then assign each worker a different set of files to read.

Pre-sharding the dataset is not very flexible to changes in the number of workers, since some workers may be assigned more data than others. For more flexibility, Ray Data provides a solution for sharding the dataset at runtime.
```

File: doc/source/train/getting-started-xgboost.rst [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+------------
+
+When running distributed XGBoost training, you will need each worker to access a different shard of the dataset.
+
+
+.. testcode:: python
+    :skipif: True
+
+    def get_train_dataset(world_rank) -> xgboost.DMatrix:
+        # Define logic to get the DMatrix for each worker
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:250]:
```suggestion
    def get_train_dataset(world_rank: int) -> xgboost.DMatrix:
        # Define logic to get the DMatrix shard for this worker rank
```

File: doc/source/train/getting-started-xgboost.rst [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+        ...
+
+A common way to do this is to pre-shard the dataset, so each worker reads from a different partition of the dataset. 
+
+For more flexibility, Ray Data provides a solution for sharding the dataset at runtime.
+
+Use Ray Data to shard the dataset
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:ref:`Ray Data <data>` is a distributed data processing library that allows you to easily shard and distribute your data across multiple workers.
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:270]: link ray train ray data guide

File: doc/source/train/getting-started-xgboost.rst [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+
+:ref:`Ray Data <data>` is a distributed data processing library that allows you to easily shard and distribute your data across multiple workers. 
+
+First, load your entire dataset as a Ray Data Dataset.
+
+.. testcode:: python
+    :skipif: True
+
+    train_dataset = ray.data.from_parquet("s3://...")
+    eval_dataset = ray.data.from_parquet("s3://...")
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:278]: reword this to highlight that you should pass the directory of ALL files into the read.

Also add a comment about how many different data sources are supported, and link to the read api reference.

File: doc/source/train/getting-started-xgboost.rst [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+
+In the training function, you can access the dataset shards for this worker using :meth:`ray.train.get_dataset_shard`. Convert this into the DMatrix.
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:280]: I think we can link xgboost.DMatrix to external xgboost docs

```suggestion
In the training function, you can access the dataset shards for this worker using :meth:`ray.train.get_dataset_shard`. Convert this into the native xgboost.DMatrix.
```

File: doc/source/train/getting-started-xgboost.rst [hunk 13]
```diff
@@ -0,0 +1,429 @@
...
+    def get_dmatrix(dataset_name: str) -> xgboost.DMatrix:
+        shard = ray.train.get_dataset_shard(dataset_name)
+        df = shard.materialize().to_pandas()
+        X, y = df.drop("target", axis=1), df["target"]
+        return xgboost.DMatrix(X, label=y)
+
+    def train_func():
+        dtrain = get_dmatrix("train")
+        deval = get_dmatrix("eval")
+        ...
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:295]: for future: let's introduce a prepare_dmatrix utility API which can implement the pyarrow version of this to reduce copies.

Comment (justinvyu) [reply doc/source/train/getting-started-xgboost.rst:295]: Explain and show a diagram about the dataset being materialized at this point and being stored in the object store. 

Include details about how this is allows the dataset to be spread across multiple nodes' CPU memory.

Also mention the default 30% object store limit and how to modify this with the env variable and how to estimate how many nodes you need.

File: doc/source/train/getting-started-xgboost.rst [hunk 14]
```diff
@@ -0,0 +1,429 @@
...
+    # 4 nodes with 8 CPUs each.
+    scaling_config = ScalingConfig(num_workers=4, resources_per_worker={"CPU": 8})
+
+    # 1 node with 8 CPUs and 4 GPUs each.
+    scaling_config = ScalingConfig(num_workers=4, use_gpu=True)
+
+    # 4 nodes with 8 CPUs and 4 GPUs each.
+    scaling_config = ScalingConfig(num_workers=16, use_gpu=True)
+
+When using GPUs, you will also need to update your training function to use the GPU. This can be done by setting the `"device"` parameter as `"cuda"`.
```
Comment (justinvyu) [inline doc/source/train/getting-started-xgboost.rst:332]: link to https://xgboost.readthedocs.io/en/stable/gpu/index.html

File: doc/source/train/train.rst [hunk 15]
```diff
@@ -10,6 +10,7 @@ Ray Train: Scalable Model Training
     PyTorch Guide <getting-started-pytorch>
     PyTorch Lightning Guide <getting-started-pytorch-lightning>
     Hugging Face Transformers Guide <getting-started-transformers>
+    XGBoost Guide <getting-started-xgboost>
```
Comment (justinvyu) [inline doc/source/train/train.rst:13]: add a link to this user guide from the old user guide and mention the deprecation.

File: doc/source/train/getting-started-xgboost.rst [hunk 16]
```diff
@@ -0,0 +1,429 @@
...
+
+1. `train_func` is the Python code that executes on each distributed training worker.
+2. :class:`~ray.train.ScalingConfig` defines the number of distributed training workers and whether to use GPUs.
+3. :class:`~ray.train.xgboost.XGBoostTrainer` launches the distributed training job.
+
+Compare a XGBoost training script with and without Ray Train.
+
+.. tab-set::
+
+    .. tab-item:: XGBoost
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply doc/source/train/getting-started-xgboost.rst:41]: I want to keep it as Before then After to contextualize what is actually different

File: doc/source/train/getting-started-xgboost.rst [hunk 17]
```diff
@@ -0,0 +1,429 @@
...
+
+You can also specify the input argument for `train_func` as a dictionary via the Trainer's `train_loop_config`. For example:
+
+.. testcode:: python
+    :skipif: True
+
+    def train_func(config):
+        label_column = config["label_column"]
+        num_boost_round = config["num_boost_round"]
+        ...
```
[the remark this answers is not in the document]
Comment (matthewdeng) [reply doc/source/train/getting-started-xgboost.rst:84]: Do you mean add `config` to all other instances of the training function definition? We originally excluded from all the quickstarts to make it easier to onboard since not everyone needs it.

File: doc/source/train/getting-started-xgboost.rst (hunk 10 above)
Comment (matthewdeng) [reply doc/source/train/getting-started-xgboost.rst:176]: This is referenced at the end of the section!

**Review by gvspraveen — APPROVED** [review]

File: doc/source/train/getting-started-xgboost.rst (hunk 13 above)
Comment (matthewdeng) [reply doc/source/train/getting-started-xgboost.rst:204]: Can I add a TODO for this? This is a bit more extensive and might be helpful to have as part of a more thorough Data + Train for XGBoost/LightGBM user guide.

File: doc/source/train/getting-started-xgboost.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (justinvyu) [reply doc/source/train/getting-started-xgboost.rst:41]: I like to have the quickstart with Ray Train immediately copy pastable. The reference code above this is just a dummy skeleton, so this is really the quickstart.

File: doc/source/train/doc_code/xgboost_quickstart.py [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+import xgboost
+from sklearn.datasets import load_iris
+
+import ray.train
+from ray.train.xgboost import XGBoostTrainer, RayTrainReportCallback
+
+# 1. Load your data as a Ray Data Dataset.
+data = load_iris(as_frame=True).frame
+dataset = ray.data.from_pandas(data)
+train_dataset, eval_dataset = dataset.train_test_split(test_size=0.2)
```
Comment (justinvyu) [inline doc/source/train/doc_code/xgboost_quickstart.py:47]: Let's just host a bunch of csv/parquet files on s3 and use `read_parquet`.

Also, read separate files/folders rather than use train_test_split.

File: doc/source/train/doc_code/xgboost_quickstart.py [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+# isort: skip_file
+
+# __xgboost_start__
+import xgboost
+from sklearn.datasets import load_iris
+from sklearn.model_selection import train_test_split
+
+# 1. Load your data as an `xgboost.DMatrix`.
+data = load_iris(as_frame=True)
+train_X, eval_X, train_y, eval_y = train_test_split(
```
Comment (justinvyu) [inline doc/source/train/doc_code/xgboost_quickstart.py:11]: then, also switch this to a `pd.read_csv("s3://...")`

File: doc/source/train/doc_code/xgboost_quickstart.py [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+
+    # 4. Do distributed data-parallel training.
+    # Ray Train sets up the necessary coordinator processes and
+    # environment variables for your workers to communicate with each other.
+    bst = xgboost.train(
+        params,
+        dtrain=dtrain,
+        evals=[(deval, "validation")],
+        num_boost_round=10,
+        # Optional: Save the model checkpoint manually instead.
```
Comment (justinvyu) [inline doc/source/train/doc_code/xgboost_quickstart.py:86]:
```suggestion
        # Optional: Use the `RayTrainReportCallback` to save and report checkpoints.
```

**Review by justinvyu — APPROVED** [review]
